### PR TITLE
Add Vultr support

### PR DESCRIPTION
**Pull request by ddymko**
## Proposed Commit Message

```
    Add support for Vultr cloud.

    This PR adds in support so that cloud-init can run on instances 
    deployed on Vultr cloud. This was originally brought up in #628.

    Co-authored-by: Eric Benner <ebenner@vultr.com>
```

## Additional Context
Greets from Vultr 👋 

This PR adds in support so that cloud-init can run on instances deploy on Vultr instances. This was originally brought up in https://github.com/canonical/cloud-init/pull/628

Please let me know if there are any updates or additions that may need to be made.

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

### CODE REVIEW

**Comment by ddymko** [comment]
@smoser as per the other PR pinging you for review of this one

File: cloudinit/sources/helpers/vultr.py [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+        METADATA = v1_json
+
+        # This comes through as a string but is JSON, make a dict
+        METADATA['vendor-config'] = json.loads(METADATA['vendor-config'])
+
+    return METADATA
+
+
+# Read the system information from SMBIOS
+def get_sysinfo():
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:86]: its not a really big deal, but it sure looks to me like you're reading 4 values from dmi here, but only ever look at 'manufacturer'.  Seems like a waste.

File: cloudinit/sources/helpers/vultr.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+    return {
+        'manufacturer': dmi.read_dmi_data("system-manufacturer"),
+        'subid': dmi.read_dmi_data("system-serial-number"),
+        'product': dmi.read_dmi_data("system-product-name"),
+        'family': dmi.read_dmi_data("system-family")
+    }
+
+
+# Get kernel parameters
+def get_kernel_parameters():
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:96]: `util.get_cmdline()`

File: cloudinit/sources/helpers/vultr.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+        EHP = EphemeralDHCPv4(net.find_fallback_nic())
+        EHP.obtain_lease()
+    except (NoDHCPLeaseError) as exc:
+        LOGGER.error("DHCP failed, cannot continue. Exception: %s",
+                     exc)
+        raise
+
+
+# Close EphermalDHCP so its not left open
+def close_ephermeral():
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:49]: the spelling is wrong here `ephemeral` not `ephermeral`.

File: cloudinit/sources/helpers/vultr.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+    # Cleanup
+    EHP = None
+
+
+# Cache the metadata for optimization
+def get_metadata(params):
+    global METADATA
+
+    if not METADATA:
+        # Bring up interface in local
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:67]: use a context for this rather htan the global EHP state.

I think you're basically just doing:
```python
with EphemeralDHCPv4(connectivity_url=params['url']):
    v1 = fetch_metadata(params)
```

and that should take care of EPH entirely.

File: cloudinit/sources/helpers/vultr.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+# Get Metadata by flag
+def fetch_metadata(params):
+    req = dict(params)
+    req['url'] = "%s/v1.json" % params['url']
+
+    return read_metadata(req)
+
+
+# Convert macs to nics
+def get_interface_name(mac):
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:172]: if you're using a global to avoid re-reading, you can use `@lru_cache`

File: cloudinit/sources/helpers/vultr.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    # Add generated network parts
+    config_template['network'] = generate_network_config(config)
+
+    # Linux specific packages
+    if util.is_Linux():
+        config_template['packages'].append("ethtool")
+
+    # Define vendor script
+    vendor_script = []
+    vendor_script.append("!/bin/bash")
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:312]: *a*ppend ? 

and thats not even a proper shebang.  odd.

**Review by smoser — CHANGES_REQUESTED** [review]
Thanks for the pull request.
I have some questions/comments in line.  I guess the big things are:
 * avoid the use of globals.
 * lots of INFO that are DEBUG
 * some of the datasources have a 'main', which can be used for easy debugging. They're inconsistent, but for a developer its a useful way to poke around.  see SmartOS or GCE for examples.
 * perhaps the `vendor_script` stuff that you're doing could make use of `activate_datasource` (see 9e904bbc3336b96475bfd00fb3bf1262ae4de49f for an explanation).
 * if the instance_id is available locally anywhere (through dmi data?) then it would be great to have a check_instance_id method in the datasource.

Again, thanks for the PR.  It will be great to have proper Vultr support in cloud-init.

**Comment by ddymko** [comment]
@smoser Thanks for the review...we'll start work on updating all of these issues/concerns

**Comment by smoser** [comment]
@ddymko inspired by your work here, i dusted off my vultr accout to build a [https://github.com/cirros-dev/cirros](cirros) release.  (next you can add support to cirros for vultr :wink: )

**Comment by ddymko** [comment]
@smoser I can't take the credit...most of the work was done by @oogy and @eb3095. 

As for Cirros please let us know if there is any way we could help!

File: tools/ds-identify [hunk 7]
```diff
@@ -124,8 +124,8 @@ DI_DSNAME=""
 # this has to match the builtin list in cloud-init, it is what will
 # be searched if there is no setting found in config.
 DI_DSLIST_DEFAULT="MAAS ConfigDrive NoCloud AltCloud Azure Bigstep \
-CloudSigma CloudStack DigitalOcean AliYun Ec2 GCE OpenNebula OpenStack \
-OVF SmartOS Scaleway Hetzner IBMCloud Oracle Exoscale RbxCloud UpCloud"
+CloudSigma CloudStack DigitalOcean Vultr AliYun Ec2 GCE OpenNebula OpenStack \
+OVF SmartOS Scaleway Hetzner IBMCloud Oracle Exoscale RbxCloud upCloud"
```
Comment (smoser) [inline tools/ds-identify:128]: you broke upcloud here.

File: cloudinit/sources/DataSourceVultr.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
+        config = vultr.generate_config(BUILTIN_DS_CONFIG)
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:52]: why would you not use `self.ds_config` here ?

File: cloudinit/sources/DataSourceVultr.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        return True
+
+    # Get the metadata by flag
+    def get_metadata(self):
+        return vultr.get_cached_metadata(BUILTIN_DS_CONFIG)
+
+    # Compare subid as instance id
+    def check_instance_id(self, sys_cfg):
+        if not vultr.is_vultr():
+            return None
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:90]: this function should return a boolean.  so `return False` here.

File: cloudinit/sources/DataSourceVultr.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+        # Baremetal has no way to implement this in local
+        if vultr.is_baremetal():
+            return None
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:94]: same. False.

File: cloudinit/sources/DataSourceVultr.py [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+    # Currently unsupported
+    @property
+    def launch_index(self):
+        return None
+
+    # Write the base configs every time. These are subject to change
+    @property
+    def network_config(self):
+        config = vultr.generate_network_config(BUILTIN_DS_CONFIG)
+        config_raw = json.dumps(config)
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:108]: this seems unnecessarily loud to log the dumpped config here. but if you wanted to do it, just skip the 'config_raw'  assignment and use 'json.dumps()` and do something like:

LOG.debug("Generated Network: %s", json.dumps(config))

File: cloudinit/sources/DataSourceVultr.py [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+    import sys
+
+    if not vultr.is_vultr():
+        print("Machine is not a Vultr instance")
+        sys.exit(1)
+
+    config = vultr.generate_config(BUILTIN_DS_CONFIG)
+    sysinfo = vultr.get_sysinfo()
+
+    print(json.dumps(sysinfo, indent=1))
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:138]: util.json_dumps() will pretty-print.  also sorts keys and other htings that you probably wanted anyway.

File: cloudinit/sources/helpers/vultr.py [hunk 13]
```diff
@@ -0,0 +1,294 @@
...
+def get_metadata(params):
+    params = json.loads(params)
+
+    # Bring up interface
+    try:
+        with EphemeralDHCPv4(connectivity_url=params['url']):
+            # Fetch the metadata
+            v1 = fetch_metadata(params)
+    except (NoDHCPLeaseError) as exc:
+        LOG.error("DHCP failed, cannot continue. Exception: %s",
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:31]: did this really need a line break ? probably not.

File: cloudinit/sources/helpers/vultr.py [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+    if "vultr" in util.get_cmdline():
+        return True
+
+    return False
+
+
+def convert_to_base64(string):
+    string_bytes = string.encode('ascii')
+    b64_bytes = base64.b64encode(string_bytes)
+    return b64_bytes.decode('ascii')
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:80]: `util.b64d`

File: cloudinit/sources/helpers/vultr.py [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+    macs_to_nic = get_interface_map()
+
+    if mac not in macs_to_nic:
+        return None
+
+    return macs_to_nic.get(mac)
+
+
+# Generate network configs
+def generate_network_config(config):
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:123]: the only thing this uses 'config' for is to call get_cached_metadata().  How about just having the caller pass in the 'md' variable ?

File: cloudinit/sources/helpers/vultr.py [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+            'encoding': 'b64',
+            'content': convert_to_base64(vendor_script),
+            'owner': 'root:root',
+            'path': '/var/lib/scripts/vendor/vultr-interface-setup.sh',
+            'permissions': '0750'
+        }
+    )
+
+    # Write the startup script
+    if script and script != "echo No configured startup script":
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:279]: this is odd.  Does the metadata service actually pass that in as the null value ?

File: cloudinit/sources/helpers/vultr.py [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+    # Prepare interface 1, private
+    if len(md['interfaces']) > 1:
+        network['config'].append(generate_private_network_interface(md))
+
+    return network
+
+
+# Input Metadata and generate public network config part
+def generate_public_network_interface(md):
+    interface_name = get_interface_name(md['interfaces'][0]['mac'])
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:151]: Am I right that this only ever uses md['interfaces'] ?  If so, you can have the one caller pass that in.

File: cloudinit/sources/DataSourceVultr.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+        subid = vultr.get_sysinfo()['subid']
+        return sources.instance_id_matches_system_uuid(subid)
+
+    # Currently unsupported
+    @property
+    def launch_index(self):
+        return None
+
+    # Write the base configs every time. These are subject to change
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:104]: i think this comment is out of place.

**Review by smoser — CHANGES_REQUESTED** [review]
Hi... I want to first say that this is really coming along well.  I don't want to discourage you with my slew of comments.  I'm just interested in cloud-init being consistent in itself and readable and maintainable.

Thanks so much, please feel free to ask for clarification.  #cloud-init is a good place for more "real time" conversation. If you dont find me available another developer might be able to help.

**Comment by ddymko** [comment]
@smoser 

We appreciate you taking the time to do a thorough review of our provider integration. It's beneficial for everyone to make sure the changes we submit, are like you said, consistent and maintainable like the rest of cloud-init.

File: cloudinit/sources/helpers/vultr.py (hunk 16 above)
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:279]: Yeah, thats what is currently passed if there is no startup script.

File: cloudinit/sources/helpers/vultr.py [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+from cloudinit import net
+from cloudinit.net.dhcp import EphemeralDHCPv4, NoDHCPLeaseError
+from functools import lru_cache
+
+# Get LOG
+LOG = log.getLogger(__name__)
+
+
+@lru_cache()
+def get_metadata(params):
```
[the remark this answers is not in the document]
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:22]: I ended up having to do this because lr_cache throws not hashable errors when a dict is either passed or returned. Figured this was a little more readable then the solutions I was seeing about getting around it.

https://stackoverflow.com/questions/6358481/using-functools-lru-cache-with-dictionary-arguments

Up to you on how you want me to go about this.

File: cloudinit/sources/helpers/vultr.py [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+                name = netcfg['name']
+                command = "ethtool -L %s combined $(nproc --all)" % name
+                vendor_script = '%s\n%s' % (vendor_script, command)
+
+    # Add write_files if it is not present in the template
+    if 'write_files' not in config_template.keys():
+        config_template['write_files'] = []
+
+    # Add vendor script to config
+    config_template['write_files'].append(
```
[the remark this answers is not in the document]
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:268]: Is there any documentation for this? I can't seem to even find a reference for it.

File: cloudinit/sources/helpers/vultr.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:22]: This may not matter in light of other changes.

File: cloudinit/sources/helpers/vultr.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:268]: Sorry, but I'm going to have ask for verification on this. I couldn't find any examples in other datasources as well. I tried implementing it based on your first example and it seems util.py is trying to pass vendor_data_raw as a string to write_file even if it is a list.

```
2021-03-10 05:00:19,275 - util.py[WARNING]: failed stage init
2021-03-10 05:00:19,275 - util.py[DEBUG]: failed stage init
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/cmd/main.py", line 652, in status_wrapper
    ret = functor(name, args)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/cmd/main.py", line 376, in main_init
    init.update()
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/stages.py", line 368, in update
    'vendordata')
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/stages.py", line 398, in _store_rawdata
    util.write_file(self._get_ipath('%s_raw' % datasource), data, 0o600)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/util.py", line 1837, in write_file
    content = encode_text(content)
  File "/usr/local/lib/python3.6/site-packages/cloud_init-21.1-py3.6.egg/cloudinit/util.py", line 117, in encode_text
    return text.encode(encoding)
AttributeError: 'list' object has no attribute 'encode'`
```

```
    self.vendordata_raw = []
    for uscript in user_scripts:
        self.vendordata_raw.append(uscript)
    self.vendordata_raw.append("#cloud-config\n%s" % json.dumps(config))`
```

I've tested the package and the scripts are ran and the config is applied so it seems like this is intended to work.

Not sure if the example you provided is not quite right, this is a bug, or I simply misunderstood or did something wrong in my implementation. I went ahead and made an adjustment that checks for lists and converts it to a string and passes that in case this is indeed a bug. Seems to work with it. PR linked bellow, if this is the case and this is an appropriate fix. If it's not a bug or I did something wrong I can just close it and rewrite this portion another way, just need clarification on what I am doing wrong here if you don't mind.

https://github.com/canonical/cloud-init/pull/837

File: cloudinit/sources/DataSourceVultr.py [hunk 21]
```diff
@@ -0,0 +1,136 @@
...
+
+        # Default hostname is "vultr"
+        if self.metadata['local-hostname'] == "":
+            self.metadata['local-hostname'] = "vultr"
+
+        self.metadata['public-keys'] = md["public-keys"].splitlines()
+        self.userdata_raw = md["user-data"]
+        if self.userdata_raw == "":
+            self.userdata_raw = None
+        self.vendordata_raw = "#cloud-config\n%s" % json.dumps(config)
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:72]: yeah... so as i read/thought about this more, this should be bytes i think.  sorry that you exposed some inconsistency in cloud-init.

Comment (eb3095) [reply cloudinit/sources/DataSourceVultr.py:72]: So to confirm you want me to change this to bytes and close the following PR?

https://github.com/canonical/cloud-init/pull/837

File: cloudinit/sources/helpers/vultr.py [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+# Confirm is Vultr
+def is_vultr():
+    # VC2, VDC, and HFC use DMI
+    sysinfo = get_sysinfo()
+
+    if sysinfo['manufacturer'] == "Vultr":
+        return True
+
+    # Baremetal requires a kernel parameter
+    if "vultr" in util.get_cmdline():
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:64]: **need to change** (but trivial)

```python
if "vultr" in util.get_cmdline().split()
```

that'll make sure you don't match (/me tries to think of a word with 'vultr' in it)..... "covultronic" :wink:.

A bit manufactured, but might as well be safe.

File: cloudinit/sources/DataSourceVultr.py [hunk 23]
```diff
@@ -0,0 +1,153 @@
...
+
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:51]: **strongly suggest changing**

the less you do here (in `_get_data`) the easier testing is on your part.  the datasource is such a pain to instantiate in testing, so doing something like this is easier:

```
dsdata = get_datasource_data(vultr.generate_config(self.metadata())
self.metadata_full = dsdata['full']
self.metadata = dsdata['metadata']
self.userdata_raw = dsdata['user-data']
self.vendordata_raw = dsdata['vendor-data']
```

Then you can write `get_datasource_data()` in a way that is easy to  mock.  Its easy to test it with different metadata values, because you passed the result of the problematic 'get_metadata()' in.

In short, I'm saying make generate_config do more so that the result of calling it is really easy for the datasource to use.

File: cloudinit/sources/DataSourceVultr.py [hunk 24]
```diff
@@ -0,0 +1,153 @@
...
+        return sources.instance_id_matches_system_uuid(subid)
+
+    # Currently unsupported
+    @property
+    def launch_index(self):
+        return None
+
+    @property
+    def network_config(self):
+        md = self.get_metadata()
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:119]: **need to change**

_get_metadata is guaranteed to have been called when network_config is called.

no need to re-call get_metadata()  you can see other datasource examples, they mainly use the cached value and then call a converter functio to turn it into what cloud-init uses.

File: cloudinit/sources/helpers/vultr.py [hunk 25]
```diff
@@ -0,0 +1,254 @@
...
+
+    if mac not in macs_to_nic:
+        return None
+
+    return macs_to_nic.get(mac)
+
+
+# Generate network configs
+def generate_network_config(md):
+    network = {
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:104]: **strongly suggest changing**

I think i'd rather this just take 'interfaces'... as that is actually all it uses.

Ie, its just kind of strange to have a function take a dictionary and then use only one value in the dictionary.

```python
func foo(mydata):
    bar = mydata['what-i-wanted']
    ... use bar here but no other part of 'mydata' and do some stuff...
```

Rather just make the caller call `foo(md['what-i-wanted'])`

File: cloudinit/sources/helpers/vultr.py [hunk 26]
```diff
@@ -0,0 +1,254 @@
...
+def generate_config(md):
+    # Create vendor config
+    config_template = copy.deepcopy(md['vendor-config'])
+
+    # Add generated network parts
+    config_template['network'] = generate_network_config(md)
+
+    # Linux specific packages
+    if util.is_Linux():
+        config_template['packages'].append("ethtool")
```
Comment (smoser) [inline cloudinit/sources/helpers/vultr.py:220]: this is less than ideal.  the reason is the way that cloud-init merges user-data and vendor-data.
if the user provides 'packages' in their user-data (and many users *will* provide packages).. then they'll trump your list here :cry: .

I don't have a really good solution for this, though.

so just file this under the "remember this" bucket.

**Review by smoser — CHANGES_REQUESTED** [review]
I hope you don't think i'm just putting you through the ringer... you really have done a good job.

I have some comments inline. If you really want to ignore the "strongly suggest changing", then you can.  Please fix the "need to change".

And then I'd  like to get this landed.

Thanks!

File: cloudinit/sources/DataSourceVultr.py (hunk 24 above)
Comment (eb3095) [reply cloudinit/sources/DataSourceVultr.py:119]: This is cached via lrucache. Did you want me to cache it further?

File: cloudinit/sources/helpers/vultr.py (hunk 26 above)
Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:220]: I have several options I can go with here.

1. Install it inside the vendor script
2. Host it and pull down the individual binary for use
3. Base64 blob somewhere and then that in the config

Which do you feel would be the most appropriate solution here? I am hesitant to ignore another major requirement of ours, especially related to networking performance.

Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:220]: @smoser Hate to bug you, but still need some input on how to best go about this.

Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:220]: Eh sorry to bother you, we ended up going another route here for this.

Comment (smoser) [reply cloudinit/sources/helpers/vultr.py:220]: @eb3095 what did that mean?  Another route as in "drop this PR"? or just wrt startup script.

Comment (eb3095) [reply cloudinit/sources/helpers/vultr.py:220]: > @eb3095 what did that mean? Another route as in "drop this PR"? or just wrt startup script.

No no no, just this specific portion. We ended up moving this to our end so we can better evaluate our options for now and removed it from the code. We will have another update to this soon.

**Comment by ddymko** [comment]
@eb3095 merged your latest changes

**Comment by eb3095** [comment]
@smoser when you get around to that review I had a question about vendor-data and the network config. I am currently adding the network config to the vendor-data kind of like how user-data functions. Is this required? Do I need to even do that? Seems like the networking config is pulled in a separate function.

**Comment by smoser** [comment]
> @smoser when you get around to that review I had a question about vendor-data and the network config. I am currently adding the network config to the vendor-data kind of like how user-data functions. Is this required? Do I need to even do that? Seems like the networking config is pulled in a separate function.

If I understand the question correctly, then the answer is 'no'.  the only place that network_config has to be present is returned by DataSource.network_config()

File: cloudinit/sources/DataSourceVultr.py [hunk 27]
```diff
@@ -0,0 +1,155 @@
...
+            util.get_cfg_by_path(sys_cfg, ["datasource", "Vultr"], {}),
+            BUILTIN_DS_CONFIG])
+        BUILTIN_DS_CONFIG['url'] = self.ds_cfg.get(
+            'url', BUILTIN_DS_CONFIG['url'])
+        BUILTIN_DS_CONFIG['retries'] = self.ds_cfg.get(
+            'retries', BUILTIN_DS_CONFIG['retries'])
+        BUILTIN_DS_CONFIG['timeout'] = self.ds_cfg.get(
+            'timeout', BUILTIN_DS_CONFIG['timeout'])
+        BUILTIN_DS_CONFIG['wait'] = self.ds_cfg.get(
+            'wait', BUILTIN_DS_CONFIG['wait'])
```
Comment (smoser) [inline cloudinit/sources/DataSourceVultr.py:39]: I'm pretty sure these lines (32-39) are not needed.

```
>>> import cloudinit.util as util
>>> util.mergemanydict([{}, BUILTIN_DS_CONFIG])
{'url': 'http://169.254.169.254', 'retries': 30, 'timeout': 2, 'wait': 2}

>>> util.mergemanydict([{'url': 'http://1.1.1.1/foo'}, BUILTIN_DS_CONFIG])
{'url': 'http://1.1.1.1/foo', 'retries': 30, 'timeout': 2, 'wait': 2}
```

Am I wrong?

Comment (smoser) [reply cloudinit/sources/DataSourceVultr.py:39]: yeah, i'm pretty sure they are *not* required anymore.  As you're just updating BUILTIN_DS_CONFIG.  but since we do not use that global anywhere else (except the main, which won't have executed *this* code), the lines are not needed.

i think they were left over from when you were using a global more.

Comment (eb3095) [reply cloudinit/sources/DataSourceVultr.py]: Indeed, just left over code I missed. Its removed now.

File: cloudinit/sources/helpers/vultr.py (hunk 25 above)
Comment (smoser) [reply cloudinit/sources/helpers/vultr.py:100]: > **strongly suggest changing**
> 
> I think i'd rather this just take 'interfaces'... as that is actually all it uses.
> 
> Ie, its just kind of strange to have a function take a dictionary and then use only one value in the dictionary.
> 
> ```python
> func foo(mydata):
>     bar = mydata['what-i-wanted']
>     ... use bar here but no other part of 'mydata' and do some stuff...
> ```
> 
> Rather just make the caller call `foo(md['what-i-wanted'])`

you did this, thanks.
My only nitpick now is the variable name 'md'.... its a bit confusing.  change it to 'interfaces' ?

File: cloudinit/sources/DataSourceVultr.py (hunk 23 above)
Comment (smoser) [reply cloudinit/sources/DataSourceVultr.py:41]: you did that too, thanks.

File: cloudinit/sources/helpers/vultr.py (hunk 25 above)
Comment (smoser) [reply cloudinit/sources/helpers/vultr.py:100]: thanks.

**Review by smoser — APPROVED** [review]

**Review by smoser — CHANGES_REQUESTED** [review]
OK. last request.

Can you squash everything and commit with the commit message in the PR?

I'm requesting that for 2 reasons:
a.) so i dont mess it up.
b.) I went to do this and saw:


    Co-authored-by: Eric Benner <ebenner@vultr.com>
    Co-authored-by: eb3095 <45504889+eb3095@users.noreply.github.com>

We'd like to have real human names in 'Author' and any other such references.

So in my head, ideally the commit would look like the following, but with
@ddymko 's  `@vultr` email address if you could.

```
Author: David Dymko <dymkod@gmail.com>
Date: Some Date

    Add support for Vultr cloud.

    This PR adds in support so that cloud-init can run on instances 
    deployed on Vultr cloud. This was originally brought up in #628.

    Co-authored-by: Eric Benner <erics-vultr@>
```

**Comment by ddymko** [comment]
@smoser I can do that but wouldn't be it be easier to do that from github to squash and merge and update the message?

**Comment by smoser** [comment]
> @smoser I can do that but wouldn't be it be easier to do that from github to squash and merge and update the message?
> 

well 2 things:
a.) when i hit 'squash and merge', i can't change the 'author'
b.) If you update the co-authored-by, i assume you will do it right (and have Eric's permission to do so).  If I do it... who knows.

> I also don't use `@vultr` email for github that won't be a problem will it?

i don't think so... although i guess you might lose the link to you if it does that based on known email addresses. I'm ok with you keeping the @gmail if you want.

**Comment by ddymko** [comment]
Yeah, I don't mind using the `@gmail`

As for the proposed commit message. I updated the message here:
https://github.com/canonical/cloud-init/pull/827#issue-583442272

I'll work on cleaning up the commits into a single one

**Review by smoser — APPROVED** [review]
